V2 React Paragraphs

pookmish opened the pull request:
# READY FOR REVIEW

# Summary
- Upgraded a whole lot of things to make rows their own entity.
- This work can be merged in separately from the associated PRs. Since this is using a major release, it will require the upgrade in the respective PRs after the merge.

# Need Review By (Date)
- 8/31

# Urgency
- medium

# Steps to Test
1. do a site install
1. checkout the branch at https://github.com/SU-SWS/acsf-cardinalsites/pull/185
1. run database updates & config imports
1. verify the UI still works the same and all the content is the same.

# See Also
- https://github.com/SU-SWS/stanford_profile_helper/pull/18
- https://github.com/SU-SWS/stanford_profile/pull/275

pookmish committed:
created entity types and bundles to start the upgrade to 2.x

pookmish committed:
further progress on v2

pookmish committed:
need to save the row field data

pookmish committed:
possibly full functioning now

pookmish committed:
some cleanup

pookmish committed:
alot of refactoring, new items are created but not editable

pookmish committed:
functional saveing

pookmish committed:
added some comments and docblocks

pookmish committed:
field preprocessing with data attributes to get styles to apply

pookmish committed:
started on the update process

pookmish committed:
created rest resource config

pookmish committed:
Merge branch '2.x-upgrade-start' of https://github.com/SU-SWS/react_paragraphs into 2.x-upgrade-start

pookmish committed:
Fixed empty help text

pookmish committed:
Renamed alot of things and inherit more of paragraphs entities

pookmish committed:
fixed some of the type form logic

pookmish committed:
reverted the libraries, this will be fixed later

pookmish committed:
fixed a couple bugs with update

pookmish committed:
Merge branch '8.x-1.x' into 2.x-upgrade-start

pookmish committed:
Merge branch '8.x-1.x' into 2.x-upgrade-start

pookmish committed:
Merge branch '8.x-1.x' into 2.x-upgrade-start

pookmish committed:
Cleanup and adjust row actions to a menu

pookmish committed:
Fixed link widget

pookmish committed:
recompiled

pookmish committed:
Merge branch '8.x-2.x' into 2.x-upgrade-start

pookmish committed:
Fixed some tests

pookmish committed:
more test

pookmish committed:
added row type form test

pookmish committed:
removed unused use statement

pookmish committed:
cc cleanup

pookmish committed:
added delete form test

pookmish committed:
Dont mark deprecated since it will break tests

pookmish committed:
updated key value storage with the new field types

pookmish committed:
Trigger tests

pookmish committed:
fixed the link widget with no values

sherakama commented on .gitignore:
```diff
@@ -1,4 +1,5 @@
 # See https://help.github.com/articles/ignoring-files/ for more about ignoring files.
+local.*
```
eh?

pookmish replied:
I'm not sure where that came from. its probably from an attempt to make the dev live reload work better.

sherakama commented on config/optional/rest.resource.entity.paragraphs_row.yml:
```diff
@@ -0,0 +1,17 @@
+langcode: en
```
Cool

sherakama replied:
Not sure this should be optional after reading through the rest of the code. Arguments for leaving it here?

pookmish replied:
meh. :shrug:

sherakama replied:
You have code that is reliant on this config so it isn't really optional in that case. I would suspect you do want to have conflicts show up with other modules if they implement the same config. It would be more clear to whoever wins that way.

pookmish replied:
:tomato: :tomato: 
:potato: :potato:

sherakama commented on config/schema/paragraphs_row_type.schema.yml:
```diff
@@ -0,0 +1,12 @@
+react_paragraphs.paragraphs_row_type.*:
+  type: config_entity
+  label: 'Paragraphs Row type config'
+  mapping:
```
No revision ID?

pookmish replied:
row types dont have revisions. its the same as a node type or paragraph type.

sherakama commented on js/build/index.html:
```diff
@@ -74,6 +77,7 @@
 
 </style>
 
+<script>const localBaseDomain = 'http://web.drupal8.loc/';</script>
```
Are you sure you want to keep this in here?

sherakama replied:
Or perhaps transition this file into an example file and gitignore it.

pookmish replied:
yeah that file is just for live dev reload. I'd like to find a better way to do local settings. but its not required for the drupal integration

sherakama commented on js/src/Components/Atoms/RowActions.js:
```diff
@@ -0,0 +1,79 @@
+import {WidgetContext} from "../../Contexts/WidgetManager";
+import {RowForm} from "../RowForm";
+import {ConfirmDialog} from "./ConfirmDialog";
+import Menu from '@material-ui/core/Menu';
+import MenuItem from '@material-ui/core/MenuItem';
```
Lots of goodies from MUI

pookmish replied:
`MUI` better

sherakama commented on js/src/Components/Widgets/LinkWidget.js:
```diff
@@ -1,26 +1,27 @@
-  if (initialCondition === undefined || initialCondition.length === 0){
-    initialCondition = [];
-    initialCondition.push(emptyLinkValue);
+  try {
+    // Checking for the length of 0 will catch any null, non-arrays, or empty
+    // arrays. This ensures the initial state is always in a good structure.
+    if (initialCondition.length === 0) {
+      throw 'Default condition failed';
+    }
+  }
+  catch (e) {
+    initialCondition = [{uri: '', title: ''}];
   }
```
Why wrap this in a try statement?

pookmish replied:
its just an easy way to skip checking all the different types. and if the `initialCondition` is actually an object instead of an array it helps since `typeof {} == typeof []` is `true` stupid javascript

pookmish replied:
I can do `if (typeof initialCondition === 'object'){...}` since it could be an array or a true object. kinda tricky way to get around all the condition checking.

sherakama replied:
Ah I get you. It just reads funny as you catch your own throw. There is probably a strategy out there for this pattern but I don't know what it is.

sherakama commented on react_paragraphs.info.yml:
```diff
@@ -1,9 +1,9 @@
 dependencies:
-    - 'drupal:rest'
-    - 'paragraphs:paragraphs'
+  - drupal:rest
```
You have the config item in the optional folder. Do you want to move the config file or remove this dependency?

pookmish replied:
I put it in the `optional` because other modules could enable similar rest configs and it prevents conflicts. :shrug:

sherakama commented on templates/paragraph_row.html.twig:
```diff
@@ -0,0 +1,22 @@
+<div{{ attributes.addClass('paragraph_row') }}>
+  {% if content %}
+    {{- content -}}
+  {% endif %}
```
Don't really need the if statement here. If the content variable is empty it will render empty.

sherakama replied:
Or, option b, if there is no content, instead of rendering an empty div, you could move the if statement to wrap the div so nothing gets rendered instead.

sherakama commented on src/ReactParagraphsServiceProvider.php:
```diff
@@ -1,35 +0,0 @@
-<?php
```
We should probably mark a bunch of the stuff in 1.x as deprecated. Or in second thought, as it is just us that is using this module, forget about it.

pookmish replied:
I started with adding `@deprecated` but that blew up the stack tests since it runs deprecation checking.

sherakama replied:
ungh. So how do we deprecate our own stuff then? We should collect the heads and come up with a deprecation strategy.

pookmish replied:
yeah i dont know the best approach there either. the `drupal-check` doesn't provide much in terms of options

sherakama commented on src/Plugin/paragraphs/Behavior/ReactBehavior.php:
```diff
@@ -0,0 +1,29 @@
+ */
+class ReactBehavior extends ParagraphsBehaviorBase {
+
+  /**
+   * {@inheritDoc}
+   *
+   * @codeCoverageIgnore Nothing to review.
+   */
+  public function view(array &$build, Paragraph $paragraph, EntityViewDisplayInterface $display, $view_mode) {
+    // Nothing to modify on the paragraph itself.
+  }
```
Can you remove this method, or whole class, as it isn't doing anything, or is it required by the interface?

pookmish replied:
its required by the interface and we need the class to register the behavior plugin.

pookmish replied:
Testing this out now... and it appears we don't actually need a behavior plugin at all to set a behavior on the paragraph/row.

sherakama commented on src/Plugin/Field/FieldWidget/ReactParagraphsWidgetBase.php:
```diff
@@ -59,19 +70,131 @@ public static function create(ContainerInterface $container, array $configuratio
+    $field_manager = \Drupal::service('entity_field.manager');
+    $fields = $field_manager->getFieldDefinitions('paragraph_row', reset($target_bundles));
+
+    // Grab the first field that is a paragraphs reference field. There should
+    // never be more than one. Maybe we can provide configurable options later.
```
Hmm. It may be interesting later to explore different field types like an entity reference field where the row would be a collection of other content somewhere in the site rather than paragraph types.

sherakama commented on src/Plugin/Field/FieldWidget/ReactParagraphsWidgetBase.php:
```diff
@@ -59,19 +70,131 @@ public static function create(ContainerInterface $container, array $configuratio
+   * @return \Drupal\field\FieldConfigInterface
+   *   Field on the paragraphs row.
+   */
+  protected static function getRowItemsField(FieldDefinitionInterface $field_definition) {
```
Should this throw an error if it cannot find the field?

pookmish replied:
:shrug: it could but if we don't have try catches for that it'll blow up anyways. we could improve the error handling later too.

sherakama commented on src/Plugin/Field/FieldWidget/ReactParagraphsWidgetBase.php:
```diff
@@ -59,19 +70,131 @@ public static function create(ContainerInterface $container, array $configuratio
+    // one bundle chosen.
+    if (
+      $field_definition->getSetting('handler') == 'default:paragraph_row' &&
+      count($field_definition->getSetting('handler_settings')['target_bundles']) == 1
```
Why constrain to only one bundle? I know there is only one right now but I am asking for future us.

pookmish replied:
just reduced scope for now.

sherakama commented on src/Plugin/Field/FieldWidget/ReactParagraphs.php:
```diff
@@ -146,179 +90,240 @@ public function formElement(FieldItemListInterface $items, $delta, array $elemen
 
-    $this->paragraphs[$item_data['id']] = $entity;
-    return $this->paragraphs[$item_data['id']];
+    // TODO: Handle validation.
```
Todo. Any tickets or something that would bring you back to this?

pookmish replied:
I haven't created tickets yet. we don't have any validation in v1 either so future improvement

sherakama commented on src/Plugin/Field/FieldType/ReactParagraphs.php:
```diff
@@ -3,18 +3,23 @@
 /**
  * Plugin implementation of the 'react_paragraphs' field type.
  *
+ * This plugin is deprecated and is only used for upgrade process. It will be
+ * Marked as deprecated and removed in a later release.
```
If it is going into version 2.x it should be marked deprecated in 2.x and removed in 3.x.

pookmish replied:
I started with adding `@deprecated` but that blew up the stack tests since it runs deprecation checking.

sherakama commented on src/Plugin/Field/FieldFormatter/ReactParagraphs.php:
```diff
@@ -2,133 +2,23 @@
 /**
  * Plugin implementation of the 'react_paragraphs' formatter.
  *
+ * This plugin is deprecated and is only used for upgrade process. It will be
+ * Marked as deprecated and removed in a later release.
```
Can we put a deprecation notice in the constructor or somewhere?

pookmish replied:
I started with adding `@deprecated` but that blew up the stack tests since it runs deprecation checking.

sherakama commented on src/ParagraphsRowTypeListBuilder.php:
```diff
@@ -0,0 +1,32 @@
+  public function buildRow(EntityInterface $entity) {
+    $row['label'] = $entity->label();
+    $row['id'] = $entity->id();
+    // You probably want a few more properties here...
```
Eh?

pookmish replied:
lol. thats left over from `drupal-console`

sherakama commented on src/ParagraphsRowTypeHtmlRouteProvider.php:
```diff
@@ -0,0 +1,15 @@
+use Drupal\Core\Entity\Routing\AdminHtmlRouteProvider;
+
+/**
+ * Provides routes for Paragraphs Row type entities.
```
Are these really needed? What purpose does having a route for these have?

pookmish replied:
Not 100% sure. they were generated by `drupal-console`

pookmish replied:
Since the class is empty, i can adjust the entity definition and remove this file entirely

sherakama commented on src/ParagraphsRowAccessControlHandler.php:
```diff
@@ -0,0 +1,24 @@
+   * {@inheritDoc}
+   */
+  protected function checkAccess(EntityInterface $entity, $operation, AccountInterface $account) {
+    return AccessResult::allowed();
```
Leaving it up to the parent entity to guard access? If the entity has a route wouldn't that leave it exposed?

pookmish replied:
The entity doesn't have any routes for editing/deleting/viewing individual entities so I don't see any harm at this time

sherakama commented on src/Form/ParagraphsRowTypeForm.php:
```diff
@@ -0,0 +1,46 @@
+    /** @var \Drupal\Core\Url $redirect */
+    $redirect = $form_state->getRedirect();
+
+    // Change which page the form redirects after submit since we don't wont to
+    // redirect to a paragraph entity area.
+    switch ($redirect->getRouteName()) {
+      case 'entity.paragraphs_type.collection':
+        $form_state->setRedirect('entity.paragraphs_row_type.collection');
+        break;
+
+      case 'entity.paragraph.field_ui_fields':
+        $route_info = FieldUI::getOverviewRouteInfo('paragraph_row', $this->entity->id());
+        $form_state->setRedirectUrl($route_info);
+        break;
+    }
+  }
```
Is this because the paragraph types try to redirect to somewhere else?

pookmish replied:
yeah since the form extends the from from the paragraphs module, we gotta switch it up.

sherakama commented on src/Entity/ParagraphsRowType.php:
```diff
@@ -0,0 +1,44 @@
+ *       "edit" = "Drupal\react_paragraphs\Form\ParagraphsRowTypeForm",
+ *       "delete" = "Drupal\react_paragraphs\Form\ParagraphsRowTypeDeleteForm"
+ *     },
+ *     "route_provider" = {
+ *       "html" = "Drupal\react_paragraphs\ParagraphsRowTypeHtmlRouteProvider",
+ *     },
```
Are routes needed?

pookmish replied:
not outside of the what the route provider provides

sherakama commented on src/Entity/ParagraphsRowType.php:
```diff
@@ -0,0 +1,44 @@
+ *     },
+ *   },
+ *   config_prefix = "paragraphs_row_type",
+ *   admin_permission = "administer site configuration",
```
Perhaps this should be its own permission

sherakama replied:
```suggestion
 *   admin_permission = "administer paragraphs row entities",
```

pookmish replied:
indeed. i'll need to add that to the permissions yml too.

sherakama commented on react_paragraphs.permissions.yml:
```diff
@@ -0,0 +1,31 @@
+edit paragraphs row entities:
+  title: 'Edit Paragraphs Row entities'
+
+view published paragraphs row entities:
+  title: 'View published Paragraphs Row entities'
+
+view unpublished paragraphs row entities:
+  title: 'View unpublished Paragraphs Row entities'
```
Are you able to unpublish a row?

pookmish replied:
not in the UI at this time.

pookmish replied:
I've removed these permissions from the file since they are currently unused

sherakama commented on react_paragraphs.module:
```diff
@@ -9,9 +9,56 @@
+ * Implements hook_preprocess_HOOK().
+ */
+function react_paragraphs_preprocess_field__react_paragraph_row(&$variables) {
+  $variables['attributes']['class'][] = 'react-paragraphs-row ';
```
```suggestion
  $variables['attributes']['class'][] = 'react-paragraphs-row';
```

sherakama commented on react_paragraphs.install:
```diff
@@ -0,0 +1,368 @@
+ * Create new rest resource.
+ */
+function react_paragraphs_update_8210() {
+  RestResourceConfig::create([
+    'id' => 'entity.paragraph_row',
+    'plugin_id' => 'entity:paragraph_row',
+    'granularity' => RestResourceConfigInterface::RESOURCE_GRANULARITY,
+    'configuration' => [
+      'methods' => ['GET'],
+      'formats' => ['json'],
+      'authentication' => ['cookie'],
+    ],
+  ])->save();
```
Should this be optional?

pookmish replied:
No it's needed for the functionality. the config is in `config/optional` since other modules/sites might have it already enabled.

sherakama commented on react_paragraphs.install:
```diff
@@ -0,0 +1,368 @@
+/**
+ * Create new rest resource.
+ */
+function react_paragraphs_update_8210() {
```
What happened to hook_update_8208, hook_update_8209?

sherakama commented on react_paragraphs.install:
```diff
@@ -0,0 +1,368 @@
+      $table_data = [];
+
+      foreach ($tables as $table) {
+        $database->schema()->dropField($table, "{$field_name}_settings");
+
+        $query = $database->select($table, 't')->fields('t')->execute();
+        while ($row = $query->fetchAssoc()) {
+          $table_data[$table][] = $row;
+        }
+        $database->truncate($table)->execute();
```
It would be nice to leave a warning in the docblock that this update is destructive and you highly recommend testing and having backups.

pookmish replied:
isn't that the idea of a "Major" release?

sherakama commented on react_paragraphs.install:
```diff
@@ -0,0 +1,368 @@
+/**
+ * Modify the field storage and field configs.
+ */
+function react_paragraphs_update_8207() {
```
What happened to 8206?

sherakama commented on react_paragraphs.install:
```diff
@@ -0,0 +1,368 @@
+/**
+ * Migrate Field data around.
+ */
+function react_paragraphs_update_8205(&$sandbox) {
```
What happened to 8204?

sherakama commented on react_paragraphs.install:
```diff
@@ -0,0 +1,368 @@
+/**
+ * Update display modes.
+ */
+function react_paragraphs_update_8203() {
```
What happened to 8202?

sherakama commented on react_paragraphs.install:
```diff
@@ -0,0 +1,368 @@
+<?php
```
![https://i.giphy.com/media/3o6ZsYzuLyRfSGX4f6/giphy.webp](https://i.giphy.com/media/3o6ZsYzuLyRfSGX4f6/giphy.webp)

sherakama commented on js/src/Contexts/WidgetManager.js:
```diff
@@ -511,23 +480,73 @@ export class WidgetManager extends Component {
       .catch(e => console.error(e));
   }
 
+  updateRowEntity(rowId, fieldName, newValues) {
```
Can you add comments to each of these functions?

pookmish replied:
Comments have been added

sherakama reviewed (changes requested):
I've got a fresh install working that I will put through the paces and then an upgrade path test.

pookmish committed:
Removed space

Co-authored-by: Sherakama <sherakama@gmail.com>

sherakama commented on js/src/Components/RowForm.js:
```diff
@@ -0,0 +1,36 @@
+import {EntityForm} from "./EntityForm";
+import {Loader} from "./Atoms/Loader";
+
+export const RowForm = ({widgetContext, open, onClose, rowId, entity, loadedEntity}) => {
```
It would be nice if this form had a message if there were no configurable options available.

sherakama replied:
Right now the edit form opens an empty modal.

pookmish replied:
Yeah I was wondering what to do with that. a basic message makes sense

pookmish replied:
I added a basic message that we can improve later if necessary

pookmish committed:
Improved & cleaned up some parts

Removed the behavior plugin since we can set and get behaviors without the plugin
Added text when an entity form has no fields
adjusted the twig template

pookmish committed:
Merge branch '2.x-upgrade-start' of https://github.com/SU-SWS/react_paragraphs into 2.x-upgrade-start

pookmish committed:
cleanup permissions and empty classes

pookmish committed:
changed update hook numbers and added comments

sherakama commented on templates/paragraph_row.html.twig:
```diff
@@ -0,0 +1,22 @@
+  {% if content %}
+    {{- content -}}
+  {% endif %}
+</div>
```
```suggestion
<div{{ attributes.addClass('paragraph_row') }}>
  {{- content -}}
</div>
```

pookmish replied:
This was changed in a recent commit https://github.com/SU-SWS/react_paragraphs/pull/59/commits/5a7df0493968f4395f897884883b9a2ca735fbcd#diff-5fac336430ff87026e14e74cd6bc0194

sherakama replied:
What history was I looking at...

pookmish commented:
![yay](https://media.tenor.com/images/863bc4209c792d8d49c47f15a3154060/tenor.gif)
